### PR TITLE
spring-boot-add-category1

### PR DESCRIPTION
**Pull request by DojiDoe**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -37,7 +37,7 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
                 .authorizeHttpRequests(
                         auth -> auth
                                 .requestMatchers("/swagger-ui/**", "/v3/api-docs/**",
-                                        "/auth/**", "/error")
+                                        "/auth/**", "/categories/**", "/error")
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/config/SecurityConfig.java]: Don’t  make categories public endpoints

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+@Tag(name = "Category management", description = "Endpoints for managing Categories")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @PostMapping
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java:35]: ResponseStatus.CREATED

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    @Operation(summary = "Create a new category")
+    public CategoryDto createCategory(@RequestBody @Valid CategoryDto categoryDto) {
+        return categoryService.save(categoryDto);
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping
+    @Operation(summary = "Get all categories", description = "Get a list of all categories")
+    public List<CategoryDto> getAll(Pageable pageable) {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java:44]: better return Page (optional)

File: src/main/java/mate/academy/bookstore/dto/book/CreateBookRequestDto.java [hunk 4]
```diff
@@ -28,4 +29,5 @@ public class CreateBookRequestDto {
     private String description;
     @NotBlank
     private String coverImage;
+    private List<Long> categoryIds;
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/dto/book/CreateBookRequestDto.java:34]: can you create a book without a category? Let’s  use NotEmpty validator

File: src/main/java/mate/academy/bookstore/dto/category/CategoryDto.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package mate.academy.bookstore.dto.category;
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+
+public record CategoryDto(
+        @NotBlank
+        @NotNull
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/dto/category/CategoryDto.java]: NotBlank includes a NotNull check

File: src/main/java/mate/academy/bookstore/model/Book.java [hunk 6]
```diff
@@ -34,4 +42,12 @@ public class Book {
     private String coverImage;
     @Column(nullable = false, columnDefinition = "TINYINT", length = 1)
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/model/Book.java]:
```suggestion
    @ManyToMany
```
It's the default fetch type for this type of relation ;)

File: src/main/java/mate/academy/bookstore/model/Book.java [hunk 7]
```diff
@@ -34,4 +42,12 @@ public class Book {
...
+    @JoinTable(
+            name = "book_category",
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/model/Book.java]:
```suggestion
            name = "books_categories",
```

File: src/main/java/mate/academy/bookstore/repository/CategoryRepository.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package mate.academy.bookstore.repository;
+
+import mate.academy.bookstore.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CategoryRepository extends JpaRepository<Category,Long> {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/repository/CategoryRepository.java]:
```suggestion
public interface CategoryRepository extends JpaRepository<Category, Long> {
```

File: src/main/java/mate/academy/bookstore/service/BookServiceImpl.java [hunk 9]
```diff
@@ -69,4 +71,12 @@ public BookDto update(UpdateBookRequestDto book) {
         bookFromDB.setCoverImage(book.getCoverImage());
         return bookMapper.toDto(bookRepository.save(bookFromDB));
     }
+
+    @Override
+    public List<BookDtoWithoutCategoryIds> findBooksByCategoryId(Long id) {
+        return bookRepository.findAllByCategoryId(id)
+                .stream()
+                .map(bookMapper::toDtoWithoutCategories)
+                .collect(Collectors.toList());
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/service/BookServiceImpl.java]:
```suggestion
                .toList();
```

File: src/main/java/mate/academy/bookstore/service/CategoryServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public CategoryDto update(Long id, CategoryDto categoryDto) {
+        Category category = categoryMapper.toModel(categoryDto);
+        category.setId(id);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public void deleteById(Long id) {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/bookstore/service/CategoryServiceImpl.java:47]: should you check that the book exists by ID before deleting it?

File: src/main/resources/db/changelog/changes/09-create-book-category-table.yaml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+            baseColumnNames: book_id
+            referencedTableName: books
+            referencedColumnNames: id
+            constraintName: fk_books_categories_books
+        - addForeignKeyConstraint:
+            baseTableName: book_category
+            baseColumnNames: category_id
+            referencedTableName: categories
+            referencedColumnNames: id
+            constraintName: fk_books_categories_categories
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/09-create-book-category-table.yaml]: add one empty line at the end of the file (check other files)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution!

File: src/main/java/mate/academy/bookstore/controller/CategoryController.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+@Tag(name = "Category management", description = "Endpoints for managing Categories")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/CategoryController.java]: wrong status

File: src/main/java/mate/academy/bookstore/service/CategoryServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        if (categoryRepository.findById(id).isPresent()) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/CategoryServiceImpl.java:48]: let's throw an exception if category not exists

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Could you please use MapStruct to update book (see screenshot) 
Screenshot 2024-12-03 at 11 52 02

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**